### PR TITLE
[Summer of code] Shrink and expand InSyncStateSet

### PR DESCRIPTION
**Pull request by hzh0425**
## What is the purpose of the change

tracking issue: https://github.com/apache/rocketmq/issues/4330

## Brief changelog

Modify the definition of syncStateSet in AutoSwitchHASerivce, and introduce the confirmOffset mechanism.

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/ha/GroupTransferService.java [hunk 1]
```diff
@@ -71,28 +74,51 @@ private void doWaitTransfer() {
...
+                            for (HAConnection conn : haService.getConnectionList()) {
+                                final AutoSwitchHAConnection autoSwitchHAConnection = (AutoSwitchHAConnection) conn;
+                                if (syncStateSet.contains(autoSwitchHAConnection.getClientAddress()) && autoSwitchHAConnection.getSlaveAckOffset() >= req.getNextOffset()) {
+                                    ackNums ++;
+                                }
+                                if (ackNums >= syncStateSet.size()) {
+                                    transferOK = true;
+                                    break;
+                                }
+                            }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/ha/GroupTransferService.java:108]: 这里的ackNums没有把master自己算进去。

File: store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java [hunk 2]
```diff
@@ -118,28 +134,93 @@ public boolean changeToSlave(String newMasterAddr, int newMasterEpoch, Long slav
...
+            for (Consumer<Set<String>> listener : syncStateSetChangedListeners) {
+                listener.accept(newSyncStateSet);
+            }
+        });
+    }
+
     public void setSyncStateSet(final Set<String> syncStateSet) {
-        this.syncStateSet = new HashSet<>(syncStateSet);
+        this.syncStateSet.clear();
+        this.syncStateSet.addAll(syncStateSet);
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java:164]: 再另一个线程里面会用syncStateSet.size判断ack，但这里更新分成了两步，所以应该会并发问题

File: store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAConnection.java [hunk 3]
```diff
@@ -279,12 +296,17 @@ protected boolean processReadResult(ByteBuffer byteBufferRead) {
...
+                                    AutoSwitchHAConnection.this.slaveAckOffset = slaveMaxOffset;
                                     if (slaveRequestOffset < 0) {
                                         slaveRequestOffset = slaveMaxOffset;
                                     }
                                     LOGGER.info("slave[" + clientAddress + "] request offset " + slaveMaxOffset);
                                     byteBufferRead.position(readSocketPos);
+                                    if (slaveMaxOffset >= AutoSwitchHAConnection.this.lastMasterMaxOffset) {
+                                        AutoSwitchHAConnection.this.lastCatchUpTimeMs = Math.max(AutoSwitchHAConnection.this.lastTransferTimeMs, AutoSwitchHAConnection.this.lastCatchUpTimeMs);
+                                        AutoSwitchHAConnection.this.haService.maybeExpandInSyncStateSet(AutoSwitchHAConnection.this.slaveAddress, slaveMaxOffset);
+                                    }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAConnection.java:308]: 这里应该也有并发问题，比如slaveMaxOffset >= AutoSwitchHAConnection.this.lastMasterMaxOffset，但写那里lastTransferTimeMs已经更新了，导致lastCatchUpTimeMs错误。

Comment (hzh0425) [reply store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAConnection.java]: done

File: store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java [hunk 4]
```diff
@@ -89,12 +103,14 @@ public boolean changeToMaster(int masterEpoch) {
         this.epochCache.appendEntry(newEpochEntry);
 
         this.defaultMessageStore.recoverTopicQueueTable();
+
+        this.syncStateSet.clear();
+        this.syncStateSet.add(this.localAddress);
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAService.java:108]: 应该调用setSyncStateSet方法

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 5]
```diff
@@ -805,13 +805,17 @@ public CompletableFuture<PutMessageResult> asyncPutMessage(final MessageExtBroke
...
+            needAckNums = -1;
+        } else {
+            if (needHandleHA) {
+                int inSyncReplicas = Math.min(this.defaultMessageStore.getAliveReplicaNumInGroup(),
+                    this.defaultMessageStore.getHaService().inSyncSlaveNums(currOffset) + 1);
+                needAckNums = calcNeedAckNums(inSyncReplicas);
+                if (needAckNums > inSyncReplicas) {
+                    // Tell the producer, don't have enough slaves to handle the send request
+                    return CompletableFuture.completedFuture(new PutMessageResult(PutMessageStatus.IN_SYNC_REPLICAS_NOT_ENOUGH, null));
+                }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java:818]: 可以放在一个if else里，不需要else里面嵌套

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**